R/xl_open.R

```diff
@@ -30,42 +33,61 @@
 #' wb$add_worksheet("Cars")
 #' wb$add_data("Cars", x, startCol = 2, startRow = 3, rowNames = TRUE)
 #' xl_open(wb)
-xl_open <- function(file = NULL) {
+xl_open <- function(x, file, interactive = NA) {
+  if (!missing(file))  {
+    warning("xl_open(file = .) is deprecated.  Use xl_open(x = .) instead")
+    x <- file
+  }
+  UseMethod("xl_open")
+}
 
-  if (!interactive()) return()
+#' @rdname xl_open
+#' @export
+xl_open.wbWorkbook <- function(x, file, interactive = NA) {
+  stopifnot(R6::is.R6(x))
+  xl_open(x$clone()$save(temp_xlsx())$path, interactive = interactive)
+}
 
-  op <- openxlsx_options()
-  on.exit(options(op), add = TRUE)
+#' @rdname xl_open
+#' @export
+xl_open.default <- function(x, file, interactive = NA) {
+  stopifnot(file.exists(x))
 
-  if (is.null(file)) stop("A file has to be specified.")
+  # nocov start
+  if (is.na(interactive)) {
+    interactive <- interactive()
+  }
+  # nocov end
 
-  ## workbook handling
-  if (inherits(file, "wbWorkbook")) {
-    file <- file$save(path = temp_xlsx())$path
+  if (!isTRUE(interactive)) {
+    warning("will not open file when not interactive")
+    return()
   }
 
-  if (!file.exists(file)) stop("Non existent file or wrong path.")
+  # nocov start
 
   ## execution should be in background in order to not block R
   ## interpreter
-  file <- normalizePath(file)
+  file <- normalizePath(x, mustWork = TRUE)
   userSystem <- Sys.info()["sysname"]
 
-
-  if ("Linux" == userSystem) {
-    if (is.null(app <- unlist(options("openxlsx2.excelApp")))) {
-      app <- chooseExcelApp()
-    }
-    myCommand <- paste(app, file, "&", sep = " ")
-    system(command = myCommand)
-  } else if ("Windows" == userSystem) {
-    shell(shQuote(file, "sh"), wait = FALSE) # nolint
-  } else if ("Darwin" == userSystem) {
-    myCommand <- paste0('open ', shQuote(file))
-    system(command = myCommand)
-  } else {
-    warning("Operating system not handled.")
-  }
+  switch(
+    userSystem,
+    Linux = {
+      app <- getOption("openxlsx2.excelApp", chooseExcelApp())
+      myCommand <- paste(app, file, "&", sep = " ")
+      system(command = myCommand)
+    },
+    Windows = {
+      shell.exec(file)
```

Comment:
#nolint at the end. This is only available on windows iirc therefore the lintr barks